### ISSUE TITLE
Extended custom network (or action) per epoch (draft)

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#311]
### Current situation and problem

The current pretrain mechanism already allows to write a custom construction function like this ([full example](https://github.com/rwth-i6/returnn-experiments/blob/master/2019-asr-e2e-trafo-vs-lstm/switchboard/base2.conv2l.specaug4a.config)):
```
def custom_construction_algo(idx, net_dict):
  if idx > final_pretrain_epoch:
    return None
  ... # modify net dict
  net_dict["#config"] = {...}  # overwrite parts of the config, like LR, batching, anything
  net_dict["#config"]["train"] = ...  # you can even overwrite the dataset ...
  return net_dict

pretrain = {"construction_algo": custom_construction_algo, ...}
```

This has turned out to be really effective, and is often used.
However, it might be slightly unintuitive.
As a rather complex example, see `test_search_multi_choice_hdf_dump`.

Here I want to discuss a new generic alternative to this, with the following goals:
* Being more intuitive (as far as possible at least...).
* Being simpler / independent of the pretrain mechanism.
* Being more flexible:
  - Allow custom actions / pre-steps per epoch or so...
  - Allow to not train an epoch, or do sth else in an epoch.
   (E.g. do a forced alignment, i.e. generate alignments and store in a HDF file.)

**Important**: Keep in mind that at any state, the job could be stopped (e.g. computation time limit reached). So when you restart the job, it should be able to recover from any state.

**Important**: When we fully want to cover one full epoch, i.e. the whole dataset (e.g. for extracting alignments), we need to be careful about partition epoch. Also, we need to be careful about multi-GPU training.

Storing something in a HDF file during training is a common aspect and should be straight-forward.
E.g. there could be the following use cases:
* Store the alignment (sparse indices) for each sequence.
* Store alignment score for each sequence.
* Store loss per sequence, or per frame.
* Implement a replay buffer (for RL-like training).

---

### Proposed solution

I propose a generic function like this in the config (independent from pretrain):
```
def get_network(epoch: int, **kwargs): ...
```
* It would return the net dict, very similar to `custom_construction_algo` above. 
* `**kwargs` should be there such that it is possible to extend it later without breaking configs.
* No need for explicit `epoch0` or so. You can simply write `epoch0 = epoch - 1` if you want to have that.
* Config overwrites as before, via `net_dict["#config"] = {...}`
* Dataset overwrite as before, via `net_dict["#config"]["train"] = {...}`
* Training can be disabled, as before, via `net_dict["#trainable"] = False`
* Custom actions should be implemented within the network. E.g. a `hdf_dump` layer can be used to dump sth to HDF.
* **TODO**: Extra check that some HDF or so is complete before we continue? How?
* **TODO**: Extra check that some part (e.g. creating HDFs) covers a full epoch (all potential sub epochs via partition_epoch in the dataset)? How?
* **TODO**: What would be a good way to debug the whole construction? It can be quite annoying to start it, see that some part of it works, and then later observe that the second or third phase or so breaks.

### Example: Transducer with full-sum training, alignment extraction, frame-wise CE training

Consider our [transducer paper (A New Training Pipeline for an Improved Neural Transducer)](https://arxiv.org/abs/2005.09319) (configs [here](https://github.com/rwth-i6/returnn-experiments/tree/master/2020-rnn-transducer)).
The final best setup would first do full-sum training, then extract an alignment, then continue frame-wise cross entropy training with this fixed alignment. The full-sum training might additionally have some pretrain epochs. The continued frame-wise cross-entropy training might import previous params, or start from scratch.
These are 3 separate steps currently but it would be nice to have this in one config, in one training run.

```
_partition_epoch = 6
train = {..., "partition_epoch": _partition_epoch}
dev = {...}
eval_datasets = {"devtrain": {...}}

_range_epochs_pretrain = (0, 10)
_range_epochs_full_sum = (0, _partition_epoch * 12)
_range_epochs_align = (_range_epochs_full_sum[1], _range_epochs_full_sum[1] + _partition_epoch)

def get_network(epoch: int, **kwargs):
  epoch0 = epoch - 1  # RETURNN starts with epoch 1, but 0-indexed is easier here
  if _range_epochs_full_sum[0] <= epoch0 < _range_epochs_full_sum[1]:
    return _get_network_full_sum(epoch0=epoch0)
  elif _range_epochs_full_sum[0] <= epoch0 < _range_epochs_full_sum[1]:
    return _get_network_align(epoch0=epoch0)
  else:
    return _get_network_default(epoch0=epoch0)

def _get_network(full_sum_loss: bool = False, ce_loss: bool = False, pretrain_frac: float = 1):
  ...  # build network dict
  return net_dict

def _get_network_full_sum(epoch0: int):
  if epoch0 < _range_epochs_pretrain[0]:
    pretrain_frac = 0
  elif epoch0 < _range_epochs_pretrain[1]:
    pretrain_frac = (
      float(epoch0 - _range_epochs_pretrain[0]) /
      (_range_epochs_pretrain[1] - _range_epochs_pretrain[0]))
  else:
    pretrain_frac = 1
  net_dict = _get_network(full_sum_loss=True, pretrain_frac=pretrain_frac)
  return net_dict

def _get_network_align(epoch0: int):
  net_dict = _get_network()
  net_dict["#trainable"] = False  # disable training
  net_dict["#finish_all_data"] = True  # in case of multi-GPU training or so
  net_dict["_align_dump"] = {
    "class": "hdf_dump",
    "from": ...,  # get alignment somehow, e.g. forced_align layer
    "is_output_layer": True,
    "dump_per_run": True,
    "extend_existing_file": epoch0 % _partition_epoch > 0,
    "filename": (lambda **opts: "data/align.{dataset_name}.hdf".format(**opts)),
  }
  return net_dict

def _get_network_default(epoch0: int):
  net_dict = _get_network(ce_loss=True)
  net_dict["#config"] = {
    ...,
    "chunking": ...  # can use chunking with frame-wise training
  }
  net_dict["#config"]["train"] = _get_dataset_with_align(train, "data/align.train.hdf")
  net_dict["#config"]["dev"] = _get_dataset_with_align(dev, "data/align.dev.hdf")
  net_dict["#config"]["eval_datasets"] = {
    key: _get_dataset_with_align(value, "data/align.%s.hdf" % key) for (key, value) in eval_datasets.items()}
  return net_dict

def _get_dataset_with_align(ds, hdf_files):
  align_ds = {"class": "HDFDataset", "files": hdf_files, "unique_seq_tags": True}
  ds = {
    "class": "MetaDataset",
    "datasets": {"original": ds, "align": align_ds},
    "data_map": {
      "data": ("original", "data"),
      "alignment": ("align", "data"),
    },
    "seq_order_control_dataset": "align",  # it must support get_all_tags
  }
  return ds

```

**Comment by JackTemaki** [comment]
I thought it might be more intuitive to pass an object that will contain the necessary attributes. Something like:
```
def modify_epoch_config(EpochConfig ep_conf):
  ep_conf.network = _get_network()
  # or
  if ep_conf.epoch > 10:
    ep_conf.network = _add_xyz_feature(ep_conf.network)
  return ep_conf
```
This would also allow to access the default network directly.

Also you could then add things to it like:
```
ep_conf.copy_param_mode = ...
ep_conf.ignore_wrong_shape = ...
ep_conf.ignore_non_existing = ...
ep_conf.config['xyz'] = ...
```

Writing things like `net_dict['#config']` feels not intuitive. "Why should the config be accessible over the network?" is what you would expect people to think. Also, in your example the network dict is generated by a custom function in your config, so how would you add the "#config" element?

**Comment by albertz** [comment]
We already have support for `#config` since a while, and there are a couple of other flags as well, which are configured that way (e.g. `#trainable`). So this would not change anything. I agree this is maybe not the cleanest way. But there are also not really problems with it. In many case, the user would also not use that feature (which is anyway somewhat experimental currently), and in those cases, it is anyway much cleaner to just return the net dict (as we also do now for pretraining).

I would also ignore the `ignore_wrong_shape` and `ignore_non_existing` for now. These are maybe too specific for this particular implementation. We might later re-implement that part.

> Also, in your example the network dict is generated by a custom function in your config, so how would you add the "#config" element?

I don't quite understand. Exactly like I demonstrate in the example? Where is the problem?

**Comment by JackTemaki** [comment]
> I don't quite understand. Exactly like I demonstrate in the example? Where is the problem?

I thought you might want to enable relative changes to the original config, like `net_dict['#config']['param_xyz'] = net_config['#config']['param_xyz']/2` but there are not really use cases for this, so I guess it is not relevant.

> But there are also not really problems with it

I did not mean there are problems, I just thought passing an object is more intuitive / flexible than doing everything over the network dict. If people look at configs that only modify the actual network with the `net_dict`, they will never assume that there is more than that (= changing the network). If there is an object (e.g. EpochConfig) and you see `epoch_config.network = ...`, then you assume that there is more beside the network.

**Comment by albertz** [comment]
>  like `net_dict['#config']['param_xyz'] = net_config['#config']['param_xyz']/2`

You can already do it like this:
```
net_dict['#config']['param_xyz'] = param_xyz / 2
```

> more intuitive / flexible

It should not be more flexible. Both are equally flexible, right? Or in what way do you think one is more flexible than the other?

More intuitive, maybe. I don't like passing a mutable object where you are supposed to write into. Instead, you could return a new object of instance `EpochConfig`. I think returning a dict is actually more intuitive for the usual case where you just want to have a custom net dict -- only the config-overwrite part in it is not so intuitive. But if you are returning an object `EpochConfig`, in any case you would need to read the documentation or the code to understand what exactly you are supposed to return, or how to use that object `EpochConfig`, which will be some (small) mental overhead.

Also, for those people who already used `#config` before, there is nothing new.

Also, when you just read an existing config, I think the variant with `#config` is quite easy to understand and straight-forward. As in the example. I don't think you will read over this `#config` part and think "this will be a `#config` layer in the network, with very strange options, and `class` missing, I don't quite understand, I just ignore this".

Note that this can also be extended later. E.g. right now we just expect that it returns a dict, and we just don't do any further special handling, so all the current already implemented logic is used (which includes also `#config`). This anyway should be kept to not break old configs. However, if we feel like it, we can later quite easily extend it, and when this function returns an object like `EpochConfig` instead of a dict, do some special handling.

**Comment by albertz** [comment]
I just added an initial (untested) implementation now (via 6d5b9abbe5ae0f911ad4676f7ad36631beaca1c9).
The example above should work (unless there are some minor bugs, but should be simple to fix).

**Comment by albertz** [comment]
I think this issue here can be closed. Further improvements can be made iterative (if they are really needed).
